File: Core/services/rag_service.py
"""
RAG Service — Retrieval Augmented Generation Pipeline.

Responsibility: Orchestrate the full RAG workflow:
  1. Retrieve relevant documents from the vector store
  2. Augment the query with document context
  3. Generate a response using the LLM

This service is pure Python, async, and fully testable.
Adapted from RAG_RAT/Core/services/rag_service.py.
"""

from __future__ import annotations

import asyncio
import logging
import time as _time
from typing import Any, AsyncGenerator, Dict, List, Optional

from Core.exceptions import (
    DocumentError,
    LLMError,
    RAGError,
    ValidationError,
)

logger = logging.getLogger(__name__)

_DEFAULT_SYSTEM_PROMPT = (
    "You are a helpful assistant. Use the provided context to answer "
    "the user's question accurately.  If the context does not contain "
    "enough information, say so honestly."
)


class RAGService:
    """
    Orchestrate the RAG pipeline.

    Pure business logic — no UI dependencies.
    Combines retrieval + augmentation + generation.
    """

    def __init__(self):
        self.llm_service = None
        self.doc_service = None

    # ─── Full Pipeline ───────────────────────────────────

    async def full_rag_pipeline(
        self,
        query: str,
        provider: str,
        model: str,
        api_key: Optional[str] = None,
        top_k: int = 5,
        temperature: float = 0.7,
        max_tokens: int = 2048,
        system_prompt: Optional[str] = None,
        **kwargs: Any,
    ) -> str:
        """
        Execute the full RAG pipeline: retrieve → augment → generate.

        Returns:
            Generated response with context.

        Raises:
            ValidationError, RAGError, LLMError
        """
        self._validate_query(query)

        try:
            # Step 1: Retrieve
            logger.info(f"RAG: Retrieving for '{query[:50]}…'")
            documents = await self._retrieve_documents(query, top_k)
            if not documents:
                logger.warning("RAG: No documents found; proceeding without context")
            else:
                logger.info(f"RAG: Retrieved {len(documents)} documents")

            # Step 2: Augment
            messages = self._augment_query(query, documents, system_prompt)

            # Step 3: Generate
            logger.info("RAG: Generating response")
            response = await self._generate_response(
                messages, provider, model, api_key, temperature, max_tokens, **kwargs
            )
            logger.info(f"✓ RAG pipeline complete: {len(response)} chars")
            return response

        except (ValidationError, RAGError, LLMError):
            raise
        except Exception as exc:
            raise RAGError(f"RAG pipeline failed: {exc}", stage="full_pipeline")

    async def stream_rag_pipeline(
        self,
        query: str,
        provider: str,
        model: str,
        api_key: Optional[str] = None,
        top_k: int = 5,
        temperature: float = 0.7,
        max_tokens: int = 2048,
        system_prompt: Optional[str] = None,
        **kwargs: Any,
    ) -> AsyncGenerator[str, None]:
        """Stream RAG response tokens."""
        self._validate_query(query)
        try:
            documents = await self._retrieve_documents(query, top_k)
            messages = self._augment_query(query, documents, system_prompt)
            async for chunk in self._stream_response(
                messages, provider, model, api_key, temperature, max_tokens, **kwargs
            ):
                yield chunk
        except (ValidationError, RAGError, LLMError):
            raise
        except Exception as exc:
            raise RAGError(f"RAG stream failed: {exc}", stage="stream_pipeline")

    async def retrieve_documents(
        self, query: str, top_k: int = 5
    ) -> List[Dict[str, Any]]:
        """Public method — retrieve documents matching *query*."""
        self._validate_query(query)
        return await self._retrieve_documents(query, top_k)

    # ─── Private Helpers ─────────────────────────────────

    async def _retrieve_documents(
        self, query: str, top_k: int
    ) -> List[Dict[str, Any]]:
        """Retrieve from RAGIntegration or return empty list."""
        try:
            try:
                from rag_integration import get_rag

                rag = await get_rag()
                if rag and rag.initialized:
                    results = await rag.search_context(
                        query, top_k=top_k, score_threshold=0.25
                    )
                    return [
                        {
                            "name": r.get("source", "unknown"),
                            "content": r.get("text", ""),
                            "score": r.get("score", 0),
                        }
                        for r in results
                    ]
            except ImportError:
                logger.debug("RAGIntegration not available")
            except Exception as exc:
                logger.warning(f"RAGIntegration search failed: {exc}")
            return []
        except Exception as exc:
            raise RAGError(f"Retrieval error: {exc}", stage="retrieval")

    def _augment_query(
        self,
        query: str,
        documents: List[Dict[str, Any]],
        system_prompt: Optional[str] = None,
    ) -> List[Dict[str, str]]:
        """Augment query with document context → OpenAI-format messages."""
        try:
            sys_msg = system_prompt or _DEFAULT_SYSTEM_PROMPT
            if not documents:
                return [
                    {"role": "system", "content": sys_msg},
                    {"role": "user", "content": query},
                ]

            context = "\n\n".join(
                f"Document: {doc.get('name', 'Unknown')}\n"
                f"Content: {doc.get('content', '')[:500]}…"
                for doc in documents[:3]
            )
            augmented = (
                f"Use the following documents to answer the query:\n\n"
                f"{context}\n\nQuery: {query}"
            )
            return [
                {"role": "system", "content": sys_msg},
                {"role": "user", "content": augmented},
            ]
        except Exception as exc:
            raise RAGError(f"Augmentation failed: {exc}", stage="augmentation")

    async def _generate_response(
        self,
        messages: List[Dict[str, str]],
        provider: str,
        model: str,
        api_key: Optional[str],
        temperature: float,
        max_tokens: int,
        **kwargs: Any,
    ) -> str:
        if self.llm_service is None:
            from Core.services.llm_service import LLMService

            self.llm_service = LLMService()
        try:
            return await self.llm_service.call_llm(
                provider=provider,
                model=model,
                messages=messages,
                api_key=api_key,
                temperature=temperature,
                max_tokens=max_tokens,
                **kwargs,
            )
        except LLMError:
            raise
        except Exception as exc:
            raise RAGError(f"Generation failed: {exc}", stage="generation")

    async def _stream_response(
        self,
        messages: List[Dict[str, str]],
        provider: str,
        model: str,
        api_key: Optional[str],
        temperature: float,
        max_tokens: int,
        **kwargs: Any,
    ) -> AsyncGenerator[str, None]:
        if self.llm_service is None:
            from Core.services.llm_service import LLMService

            self.llm_service = LLMService()
        try:
            async for chunk in self.llm_service.stream_llm(
                provider=provider,
                model=model,
                messages=messages,
                api_key=api_key,
                temperature=temperature,
                max_tokens=max_tokens,
                **kwargs,
            ):
                yield chunk
        except LLMError:
            raise
        except Exception as exc:
            raise RAGError(f"Stream generation failed: {exc}", stage="generation")

    @staticmethod
    def _validate_query(query: str) -> None:
        if not query or not query.strip():
            raise ValidationError("Query cannot be empty", field="query")
        if len(query) > 100_000:
            raise ValidationError("Query too long (max 100 000 chars)", field="query")

File: rust_output/src/rag_service.rs
/// RAG Service — Retrieval Augmented Generation Pipeline.
/// 
/// Responsibility: Orchestrate the full RAG workflow:
/// 1. Retrieve relevant documents from the vector store
/// 2. Augment the query with document context
/// 3. Generate a response using the LLM
/// 
/// This service is pure Python, async, and fully testable.
/// Adapted from RAG_RAT/Core/services/rag_service::py.

use anyhow::{Result, Context};
use crate::exceptions::{LLMError, RAGError, ValidationError};
use std::collections::HashMap;
use tokio;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

pub const _DEFAULT_SYSTEM_PROMPT: &str = "You are a helpful assistant. Use the provided context to answer the user's question accurately.  If the context does not contain enough information, say so honestly.";

/// Orchestrate the RAG pipeline.
/// 
/// Pure business logic — no UI dependencies.
/// Combines retrieval + augmentation + generation.
#[derive(Debug, Clone)]
pub struct RAGService {
    pub llm_service: Option<serde_json::Value>,
    pub doc_service: Option<serde_json::Value>,
}

impl RAGService {
    pub fn new() -> Self {
        Self {
            llm_service: None,
            doc_service: None,
        }
    }
    /// Execute the full RAG pipeline: retrieve → augment → generate.
    /// 
    /// Returns:
    /// Generated response with context.
    /// 
    /// Raises:
    /// ValidationError, RAGError, LLMError
    pub async fn full_rag_pipeline(&mut self, query: String, provider: String, model: String, api_key: Option<String>, top_k: i64, temperature: f64, max_tokens: i64, system_prompt: Option<String>, kwargs: HashMap<String, Box<dyn std::any::Any>>) -> Result<String> {
        // Execute the full RAG pipeline: retrieve → augment → generate.
        // 
        // Returns:
        // Generated response with context.
        // 
        // Raises:
        // ValidationError, RAGError, LLMError
        self._validate_query(query);
        // try:
        {
            logger.info(format!("RAG: Retrieving for '{}…'", query[..50]));
            let mut documents = self._retrieve_documents(query, top_k).await;
            if !documents {
                logger.warning("RAG: No documents found; proceeding without context".to_string());
            } else {
                logger.info(format!("RAG: Retrieved {} documents", documents.len()));
            }
            let mut messages = self._augment_query(query, documents, system_prompt);
            logger.info("RAG: Generating response".to_string());
            let mut response = self._generate_response(messages, provider, model, api_key, temperature, max_tokens, /* ** */ kwargs).await;
            logger.info(format!("✓ RAG pipeline complete: {} chars", response.len()));
            response
        }
        // except (ValidationError, RAGError, LLMError) as _e:
        // except Exception as exc:
    }
    /// Stream RAG response tokens.
    pub async fn stream_rag_pipeline(&mut self, query: String, provider: String, model: String, api_key: Option<String>, top_k: i64, temperature: f64, max_tokens: i64, system_prompt: Option<String>, kwargs: HashMap<String, Box<dyn std::any::Any>>) -> Result<AsyncGenerator</* unknown */>> {
        // Stream RAG response tokens.
        self._validate_query(query);
        // try:
        {
            let mut documents = self._retrieve_documents(query, top_k).await;
            let mut messages = self._augment_query(query, documents, system_prompt);
            // async for
            while let Some(chunk) = self._stream_response(messages, provider, model, api_key, temperature, max_tokens, /* ** */ kwargs).next().await {
                /* yield chunk */;
            }
        }
        // except (ValidationError, RAGError, LLMError) as _e:
        // except Exception as exc:
    }
    /// Public method — retrieve documents matching *query*.
    pub async fn retrieve_documents(&self, query: String, top_k: i64) -> Vec<HashMap<String, Box<dyn std::any::Any>>> {
        // Public method — retrieve documents matching *query*.
        self._validate_query(query);
        self._retrieve_documents(query, top_k).await
    }
    /// Retrieve from RAGIntegration or return empty list.
    pub async fn _retrieve_documents(&self, query: String, top_k: i64) -> Result<Vec<HashMap<String, Box<dyn std::any::Any>>>> {
        // Retrieve from RAGIntegration or return empty list.
        // try:
        {
            // try:
            {
                // TODO: from rag_integration import get_rag
                let mut rag = get_rag().await;
                if (rag && rag.initialized) {
                    let mut results = rag.search_context(query, /* top_k= */ top_k, /* score_threshold= */ 0.25_f64).await;
                    results.iter().map(|r| HashMap::from([("name".to_string(), r.get(&"source".to_string()).cloned().unwrap_or("unknown".to_string())), ("content".to_string(), r.get(&"text".to_string()).cloned().unwrap_or("".to_string())), ("score".to_string(), r.get(&"score".to_string()).cloned().unwrap_or(0))])).collect::<Vec<_>>()
                }
            }
            // except ImportError as _e:
            // except Exception as exc:
            vec![]
        }
        // except Exception as exc:
    }
    /// Augment query with document context → OpenAI-format messages.
    pub fn _augment_query(&self, query: String, documents: Vec<HashMap<String, Box<dyn std::any::Any>>>, system_prompt: Option<String>) -> Result<Vec<HashMap<String, String>>> {
        // Augment query with document context → OpenAI-format messages.
        // try:
        {
            let mut sys_msg = (system_prompt || _DEFAULT_SYSTEM_PROMPT);
            if !documents {
                vec![HashMap::from([("role".to_string(), "system".to_string()), ("content".to_string(), sys_msg)]), HashMap::from([("role".to_string(), "user".to_string()), ("content".to_string(), query)])]
            }
            let mut context = documents[..3].iter().map(|doc| format!("Document: {}\nContent: {}…", doc.get(&"name".to_string()).cloned().unwrap_or("Unknown".to_string()), doc.get(&"content".to_string()).cloned().unwrap_or("".to_string())[..500])).collect::<Vec<_>>().join(&"\n\n".to_string());
            let mut augmented = format!("Use the following documents to answer the query:\n\n{}\n\nQuery: {}", context, query);
            vec![HashMap::from([("role".to_string(), "system".to_string()), ("content".to_string(), sys_msg)]), HashMap::from([("role".to_string(), "user".to_string()), ("content".to_string(), augmented)])]
        }
        // except Exception as exc:
    }
    pub async fn _generate_response(&mut self, messages: Vec<HashMap<String, String>>, provider: String, model: String, api_key: Option<String>, temperature: f64, max_tokens: i64, kwargs: HashMap<String, Box<dyn std::any::Any>>) -> Result<String> {
        if self.llm_service::is_none() {
            // TODO: from Core.services.llm_service import LLMService
            self.llm_service = LLMService();
        }
        // try:
        {
            self.llm_service::call_llm(/* provider= */ provider, /* model= */ model, /* messages= */ messages, /* api_key= */ api_key, /* temperature= */ temperature, /* max_tokens= */ max_tokens, /* ** */ kwargs).await
        }
        // except LLMError as _e:
        // except Exception as exc:
    }
    pub async fn _stream_response(&mut self, messages: Vec<HashMap<String, String>>, provider: String, model: String, api_key: Option<String>, temperature: f64, max_tokens: i64, kwargs: HashMap<String, Box<dyn std::any::Any>>) -> Result<AsyncGenerator</* unknown */>> {
        if self.llm_service::is_none() {
            // TODO: from Core.services.llm_service import LLMService
            self.llm_service = LLMService();
        }
        // try:
        {
            // async for
            while let Some(chunk) = self.llm_service::stream_llm(/* provider= */ provider, /* model= */ model, /* messages= */ messages, /* api_key= */ api_key, /* temperature= */ temperature, /* max_tokens= */ max_tokens, /* ** */ kwargs).next().await {
                /* yield chunk */;
            }
        }
        // except LLMError as _e:
        // except Exception as exc:
    }
    pub fn _validate_query(query: String) -> Result<()> {
        if (!query || !query.trim().to_string()) {
            return Err(anyhow::anyhow!("ValidationError('Query cannot be empty', field='query')"));
        }
        if query.len() > 100000 {
            return Err(anyhow::anyhow!("ValidationError('Query too long (max 100 000 chars)', field='query')"));
        }
    }
}
